### PR TITLE
feat(batch): apply row-based encoding for materialize executor and batch mode

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
close https://github.com/singularity-data/risingwave/issues/4367

### CODE REVIEW

File: src/stream/src/from_proto/batch_query.rs [hunk 1]
```diff
@@ -65,6 +65,11 @@ impl ExecutorBuilder for BatchQueryExecutorBuilder {
             .copied()
             .map(ColumnId::from)
             .collect();
+        // let column_ids = table_desc
```
Comment (hzxa21) [inline src/stream/src/from_proto/batch_query.rs:68]: remove these?

File: src/batch/src/executor/row_seq_scan.rs [hunk 2]
```diff
@@ -49,8 +49,8 @@ pub struct RowSeqScanExecutor<S: StateStore> {
 }
 
 pub enum ScanType<S: StateStore> {
-    TableScan(BatchDedupPkIter<S, CellBasedRowSerde>),
-    RangeScan(StorageTableIter<S, CellBasedRowSerde>),
+    TableScan(StorageTableIter<S, RowBasedSerde>),
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/batch/src/executor/row_seq_scan.rs:52]: With dedup fully implemented, they should also use the same iterator. For now I think there's no need to differentiate them.

File: src/stream/src/executor/batch_query.rs [hunk 3]
```diff
@@ -45,25 +41,21 @@ where
     const DEFAULT_BATCH_SIZE: usize = 100;
 
     pub fn new(
-        table: StorageTable<S, READ_ONLY>,
+        table: RowBasedStorageTable<S, READ_ONLY>,
         batch_size: Option<usize>,
         info: ExecutorInfo,
-        pk_descs: Vec<OrderedColumnDesc>,
+        _pk_descs: Vec<OrderedColumnDesc>,
```
Comment (BugenZhao) [inline src/stream/src/executor/batch_query.rs:47]: May leave a TODO here: should be removed later.

File: src/storage/src/row_serde/row_based_deserializer.rs [hunk 4]
```diff
@@ -64,11 +65,19 @@ impl RowBasedDeserializer {
...
+            row_based_deserialize_inner(&self.column_mapping.all_data_types, value.as_ref())?;
+        let output_column_ids = self
+            .column_mapping
+            .output_columns
+            .iter()
+            .map(|c| c.column_id)
+            .collect_vec();
+        let mut output_row = vec![];
+        for col_id in output_column_ids {
+            output_row.push(origin_row.0[col_id.get_id() as usize].clone());
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/src/row_serde/row_based_deserializer.rs]: Here we want to create a sub `vec` from a  complete `vec`. Fortunately😊, the elements in `Row` are `Option<ScalarImpl>`, so it's easy to use `take()` to get ownership, this is a _syntactic sugar_ in rust:
```rust
pub const fn take(&mut self) -> Option<T> {
        // FIXME replace `mem::replace` by `mem::take` when the latter is const ready
        mem::replace(self, None)
    }
```
If the element is not `Option`, avoiding `clone` is a little more complicated.

File: src/storage/src/row_serde/mod.rs [hunk 5]
```diff
@@ -96,11 +96,16 @@ pub trait RowSerialize: Clone {
 /// Record mapping from [`ColumnDesc`], [`ColumnId`], and output index of columns in a table.
 #[derive(Clone)]
 pub struct ColumnDescMapping {
+    /// output_columns are some of the columns that to be partialy scan.
     pub output_columns: Vec<ColumnDesc>,
 
-    pub id_to_column_index: HashMap<ColumnId, usize>,
+    pub ouput_id_to_index: HashMap<ColumnId, usize>,
```
Comment (BugenZhao) [inline src/storage/src/row_serde/mod.rs:102]:
```suggestion
    /// .. is the output column's column index in output row, which is used in cell-based deserialize.
    pub output_id_to_index: HashMap<ColumnId, usize>,
```

File: src/storage/src/row_serde/mod.rs [hunk 6]
```diff
@@ -96,11 +96,16 @@ pub trait RowSerialize: Clone {
...
 
+    /// all_data_types is the full row data types, which is used in row-based deserialize.
     pub all_data_types: Vec<DataType>,
+
+    /// output_index is the output column's column index in full row.
```
Comment (BugenZhao) [inline src/storage/src/row_serde/mod.rs:107]:
```suggestion
    /// output_index is the output column's column index in full row, which is used in row-based deserialize.
```

**Review by BugenZhao — APPROVED** [review]
LGTM

File: src/batch/src/executor/row_seq_scan.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/batch/src/executor/row_seq_scan.rs:52]: Will refactor this part in next PR.

**Review by hzxa21 — APPROVED** [review]
Good work! Thanks for the PR!

**Pull request merged**